### PR TITLE
[GSOC] integrated support for zvbb insn

### PR DESCRIPTION
**Pull request by SyedHassanUlHaq**

### CODE REVIEW

**Comment by ksco** [comment]
Thanks, please fix the CI.

**Comment by SyedHassanUlHaq** [comment]
> Thanks, please fix the CI.

but zvbb is a valid prefixed ISA extension for vector cyrpto.

**Comment by ksco** [comment]
I think it's the toolchain used in CI needs an upgrade, as seen from the error message.

**Comment by sequencer** [comment]
@SyedHassanUlHaq check here:
https://github.com/ksco/riscv-vector-tests/blob/bafa717d37b9bef3e80b66a50b01c22f532306bc/.github/workflows/build-and-test.yml#L47

**Comment by ksco** [comment]
> @SyedHassanUlHaq check here:
> 
> https://github.com/ksco/riscv-vector-tests/blob/bafa717d37b9bef3e80b66a50b01c22f532306bc/.github/workflows/build-and-test.yml#L47

I think it's the toolchain version that is too old, not Spike.

**Comment by SyedHassanUlHaq** [comment]
yes the Toolchain was too old with GCC 12.2, I've upgraded it for GCC 13.2.

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -37,7 +37,7 @@ jobs:
       - name: Install Toolchain
         if: steps.cache-spike.outputs.cache-hit != 'true'
         run: |
-          wget -O- -q https://github.com/riscv-collab/riscv-gnu-toolchain/releases/download/2023.07.07/riscv64-elf-ubuntu-22.04-gcc-nightly-2023.07.07-nightly.tar.gz | tar -xzf -
+          wget -O- -q wget -O- -q https://github.com/riscv-collab/riscv-gnu-toolchain/releases/download/2024.04.12/riscv64-elf-ubuntu-22.04-gcc-nightly-2024.04.12-nightly.tar.gz | tar -xzf -
```
Comment (ksco) [inline .github/workflows/build-and-test.yml:40]: `wget -O- -q` repeated twice, probably not intended.

**Comment by ksco** [comment]
CI remains unhappy, I suppose you've tested this PR locally? what's the toolchain version you have on your machine?

**Comment by SyedHassanUlHaq** [comment]
> CI remains unhappy, I suppose you've tested this PR locally? what's the toolchain version you have on your machine?

2024.04.12

**Comment by ksco** [comment]
> 2024.04.12

If that's true, CI should be passing now, no?

**Comment by SyedHassanUlHaq** [comment]
> > 2024.04.12
> 
> If that's true, CI should be passing now, no?

Screenshot from 2024-07-15 15-14-53


The CI worklow skipped jobs probably because they were cached

**Comment by ksco** [comment]
Oh indeed. I deleted the cache, let’s give it another go.

**Comment by SyedHassanUlHaq** [comment]
```
build/merger -stage1output out/v256x64machine/tests/stage1/ -stage2output out/v256x64machine/tests/stage2/ -stage2patch out/v256x64machine/patches/stage2/
fatal: wrong patch generated for file vwsll.vi-1.S, expected 554, got 211
```

I'm encountering an issue with generating the correct number of patches for the vwsll.vi instruction during the test case generation process. Despite multiple attempts to refine the configurations, the expected and actual patch counts do not match.

is it a known issue?

**Comment by ksco** [comment]
You need to debug it to see why. Start with running the test using Spike manually.

**Comment by SyedHassanUlHaq** [comment]
The compiled stage1 binaries run manually on spike successfully. I debugged the code but the error remains unresolved so basically the compiled stage1 binary of vwsll.vi gives the error ```*** FAILED *** (tohost = 668)``` when loaded on pspike. All the other instructions even those with the same instruction format works fine. I've tried debugging the generator code but couldn't find any bug that explains why the program is unable to communicate. However I haven't been able to debug pspike much because of lack of resources to understand it.

Thanks

**Comment by ksco** [comment]
You need to run the problematic program under standard Spike, not pspike, then you can use the `-d` option to debug the program.

**Comment by SyedHassanUlHaq** [comment]
> You need to run the problematic program under standard Spike, not pspike, then you can use the `-d` option to debug the program.

I did but can't find any difference in the log of a previous integrated insn and the log of a vwsll.vi

**Comment by ksco** [comment]
The difference is the problematic program exited early because of some ill-formed generated code, which is why the counts didn't match. You need to debug the program using Spike to figure out the cause and fix the codegen.

**Comment by SyedHassanUlHaq** [comment]
> The difference is the problematic program exited early because of some ill-formed generated code, which is why the counts didn't match. You need to debug the program using Spike to figure out the cause and fix the codegen.

Oh I found the problematic program, for SEW=32 spike gives illegal instruction trap at ```vwsll.vi v1, v3, 3, v0.t```
```
(spike) core   0: 0x00000000800001aa (0xd431b0d7) vwsll.vi v1, v3, 3, v0.t
core   0: exception trap_illegal_instruction, epc 0x00000000800001aa
core   0:           tval 0x00000000d431b0d7
```

**Comment by Lucas-Wye** [comment]
> > The difference is the problematic program exited early because of some ill-formed generated code, which is why the counts didn't match. You need to debug the program using Spike to figure out the cause and fix the codegen.
> 
> Oh I found the problematic program, for SEW=32 spike gives illegal instruction trap at `vwsll.vi v1, v3, 3, v0.t`
> 
> ```
> (spike) core   0: 0x00000000800001aa (0xd431b0d7) vwsll.vi v1, v3, 3, v0.t
> core   0: exception trap_illegal_instruction, epc 0x00000000800001aa
> core   0:           tval 0x00000000d431b0d7
> ```

if it is a problem from spike, pls create a issue to the repo of spike

**Comment by SyedHassanUlHaq** [comment]
> > > The difference is the problematic program exited early because of some ill-formed generated code, which is why the counts didn't match. You need to debug the program using Spike to figure out the cause and fix the codegen.
> > 
> > 
> > Oh I found the problematic program, for SEW=32 spike gives illegal instruction trap at `vwsll.vi v1, v3, 3, v0.t`
> > ```
> > (spike) core   0: 0x00000000800001aa (0xd431b0d7) vwsll.vi v1, v3, 3, v0.t
> > core   0: exception trap_illegal_instruction, epc 0x00000000800001aa
> > core   0:           tval 0x00000000d431b0d7
> > ```
> 
> if it is a problem from spike, pls create a issue to the repo of spike

No, the problem with the insn is that vd is not a multiple of EMUL. So I am modifying the generator so that it generates instructions with valid vd and vs2

File: generator/insn_vdvs2uimmvm.go [hunk 2]
```diff
@@ -29,14 +29,33 @@ func (i *Insn) genCodeVdVs2UimmVm(pos int) []string {
...
+					break
+				}
+			}else{
+				vd = 4
+				vs2 = 2
+			}
+		}else {
+			vd = int(c.LMUL1)
+			vs2 = 2*int(c.LMUL1) + int(vs2EMUL1)
+		}
```
Comment (ksco) [inline generator/insn_vdvs2uimmvm.go:57]: This needs to be more general, please refer to other `vw`-prefixed instruction generators to see how vd and vs2 are computed.

File: generator/insn_vdvs2uimmvm.go [hunk 3]
```diff
@@ -29,14 +29,33 @@ func (i *Insn) genCodeVdVs2UimmVm(pos int) []string {
...
+		vd = int(c.LMUL1)
+		vs2 = 2*int(c.LMUL1) + int(vs2EMUL1)
+
 		builder.WriteString(i.gWriteRandomData(c.LMUL1))
 		builder.WriteString(i.gLoadDataIntoRegisterGroup(vd, c.LMUL1, SEW(8)))
 
 		builder.WriteString(i.gWriteIntegerTestData(vs2EMUL1, vs2EEW, 1))
 		builder.WriteString(i.gLoadDataIntoRegisterGroup(vs2, vs2EMUL1, vs2EEW))
 
+		if i.Name == "vwsll.vi" {
```
Comment (ksco) [inline generator/insn_vdvs2uimmvm.go:42]: This needs to be more general, please refer to other `vw`-prefixed instruction generators to see how it's done.

**Review by ksco — COMMENTED** [review]
Please make the code more general, and well formatted using `gofmt`.

**Comment by ksco** [comment]
Looks good now, thank you for this!

**Pull request merged**